crates/sui-core/src/authority/submitted_transaction_cache.rs

```diff
@@ -0,0 +1,311 @@
+#[derive(Debug, Clone)]
+struct SubmissionMetadata {
+    /// Number of times this transaction has been submitted
+    submission_count: u32,
+    /// Maximum allowed submissions based on gas price amplification
+    max_allowed_submissions: u32,
+    /// List of client IP addresses that have submitted this transaction
+    submitter_client_addrs: Vec<IpAddr>,
+}
+
+impl SubmittedTransactionCache {
+    pub(crate) fn new(cache_capacity: Option<usize>) -> Self {
+        let capacity = cache_capacity
+            .and_then(NonZeroUsize::new)
+            .unwrap_or_else(|| NonZeroUsize::new(DEFAULT_CACHE_CAPACITY).unwrap());
+
+        Self {
+            inner: RwLock::new(Inner {
+                transactions: LruCache::new(capacity),
+            }),
+        }
+    }
+
+    pub(crate) fn record_submitted_tx(
+        &self,
+        digest: &TransactionDigest,
+        amplification_factor: u32,
+        submitter_client_addr: Option<IpAddr>,
+    ) {
+        let mut inner = self.inner.write();
+
+        let max_allowed_submissions = amplification_factor;
+
+        if let Some(metadata) = inner.transactions.get_mut(digest) {
+            // Track additional client addresses for resubmissions
+            if let Some(addr) = submitter_client_addr {
+                if !metadata.submitter_client_addrs.contains(&addr) {
+                    metadata.submitter_client_addrs.push(addr);
+                    debug!("Added new client address {addr} for transaction {digest}");
+                }
+            }
+            debug!("Transaction {digest} already tracked in submission cache");
+        } else {
+            // First time we're submitting this transaction, however we will wait till
+            // we see the transaction in consensus output to increment the submission count.
+            let submitter_client_addrs = submitter_client_addr.into_iter().collect();
+            let metadata = SubmissionMetadata {
+                submission_count: 0,
+                max_allowed_submissions,
+                submitter_client_addrs,
+            };
+
+            inner.transactions.put(*digest, metadata);
+
+            debug!(
+                "First submission of transaction {digest} (max_allowed: {max_allowed_submissions})",
+            );
+        }
+    }
+
+    /// Increments the submission count when we see a transaction in consensus output.
+    /// This tracks how many times the transaction has appeared in consensus (from any validator).
+    /// Returns the spam weight and list of submitter client addresses if the transaction exceeds allowed submissions.
+    pub(crate) fn increment_submission_count(
+        &self,
+        digest: &TransactionDigest,
+    ) -> Option<(Weight, Vec<IpAddr>)> {
+        let mut inner = self.inner.write();
+
+        if let Some(metadata) = inner.transactions.get_mut(digest) {
+            metadata.submission_count += 1;
+
+            if metadata.submission_count > metadata.max_allowed_submissions {
```

Comment:
I would suggest adding some high level metrics to capture the spam behaviour. As far as I can see the traffic controller doesn't allow us to "tag" the path of spam. For example maybe track number of transactions over the max allowed submissions